packages/token-generation-readmodel-checker/src/utils/utils.ts

```diff
@@ -0,0 +1,1242 @@
+
+            const purposeState = getPurposeStateFromPurposeVersions(
+              purpose.versions
+            );
+            const lastPurposeVersion = getLastPurposeVersion(purpose.versions);
+            const agreement = consumerIdEserviceIdMap.get(
+              makeGSIPKConsumerIdEServiceId({
+                consumerId: client.consumerId,
+                eserviceId: purpose.eserviceId,
+              })
+            );
+
+            if (!agreement) {
+              logger.error(
+                `token-generation-states entry with PK ${e.PK} has no agreement`
+              );
+              console.log(
+                `token-generation-states entry with PK ${e.PK} has no agreement`
+              );
+
+              return [
+                ...acc,
+                {
+                  PK: e.PK,
+                  consumerId: e.consumerId,
+                  clientKind: e.clientKind,
+                  publicKey: e.publicKey,
+                  GSIPK_clientId: e.GSIPK_clientId,
+                  GSIPK_kid: e.GSIPK_kid,
+                  GSIPK_clientId_purposeId: e.GSIPK_clientId_purposeId,
+                  GSIPK_purposeId: e.GSIPK_purposeId,
+                  purposeState: e.purposeState,
+                  purposeVersionId: e.purposeVersionId,
+                  GSIPK_eserviceId_descriptorId:
+                    e.GSIPK_eserviceId_descriptorId,
+                },
+              ];
+            }
+            const agreementItemState = agreementStateToItemState(
+              agreement.state
+            );
+
+            const eservice = eserviceIdDescriptorIdMap.get(
+              makeGSIPKEServiceIdDescriptorId({
+                eserviceId: agreement.eserviceId,
+                descriptorId: agreement.descriptorId,
+              })
+            );
+
+            if (!eservice) {
+              console.log(
+                `token-generation-states entry with PK ${e.PK} has no eservice`
+              );
+              logger.error(
+                `token-generation-states entry with PK ${e.PK} has no eservice`
+              );
+
+              return [
+                ...acc,
+                {
+                  PK: e.PK,
+                  consumerId: e.consumerId,
+                  clientKind: e.clientKind,
+                  publicKey: e.publicKey,
+                  GSIPK_clientId: e.GSIPK_clientId,
+                  GSIPK_kid: e.GSIPK_kid,
+                  GSIPK_clientId_purposeId: e.GSIPK_clientId_purposeId,
+                  GSIPK_purposeId: e.GSIPK_purposeId,
+                  purposeState: e.purposeState,
+                  purposeVersionId: e.purposeVersionId,
+                  GSIPK_eserviceId_descriptorId:
+                    e.GSIPK_eserviceId_descriptorId,
+                  agreementId: e.agreementId,
+                  agreementState: e.agreementState,
+                },
+              ];
+            }
+            const descriptor = getLastEServiceDescriptor(eservice.descriptors);
```

Comment:
this should not take the last descriptor, but the descriptor indicated in the read model agreement